### PR TITLE
Ownership-aware scheduling.

### PR DESCRIPTION
**Pull request by ryzhyk**
Background
----------

A stream in a circuit can be connected to multiple consumers.  It is
therefore generally impossible to provide each consumer with an owned
copy of the data without cloning it.  At the same time, many operators
can be more efficient when working with owned inputs.  For instance,
when computing a sum of two z-sets, if one of the input z-sets is owned
we can just add values from the other z-set to it without cloning the
first z-set.  If both inputs are owned then we additionally do not need
to clone key/value pairs when inserting them.  Furthermore, the
implementation can choose to add the contents of the smaller z-set to
the larger one.

Ownership-aware scheduling
--------------------------

To leverage such optimizations, we adopt the best-effort approach:
operators consume streaming data by-value when possible while falling
back to pass-by-reference otherwise.  In a synchronous circuit, each
operator reads its input stream precisely once in each clock cycle.  It
is therefore possible to determine the last consumer at each clock cycle
and give it the owned value from the channel.  It is furthermore
possible for the scheduler to schedule operators that strongly prefer
owned values last.

We capture ownership preferences at two levels.  First, each individual
operator that consumes one or more streams exposes its preferences on a
per-stream basis via an API method (e.g.,
`[UnaryOperator::input_preference]`).  The [`Circuit`] API allows the
circuit builder to override these preferences when instantiating an
operator, taking into account circuit topology and workload.  We express
preference as a numeric value.

These preferences are associated with each edge in the circuit graph.
The schedulers (both static and dynamic)  have been extended to
implement a limited form of ownership-aware scheduling.  They only
consider strong preferences (`OwnershipPreference::require_owned` and
stronger) and model them as hard constraints that must be satisfied for
the circuit to be schedulable.  Weaker preferences are ignored.

New operators
-------------

This commit also adds new operators needed to test ownership-aware
scheduling:
- Binary plus
- Integrator

We also start adding convenience APIs for instantiating operators using
extension traits, so that instead of
`Circuit::add_unary_operator(Inspect::new(|x| ...), stream)` one can
just write `stream.inspect(|x| ...)`.

Issues and limitations
----------------------

The above changes make the `enter` API potentially dangerous.  There is
nothing preventing the child circuit to read from the entered stream
multiple times, thus removing the value from the stream while it is
still needed by other consumers.  We will address this in an upcoming
PR.

The new scheduling policy could also do with more testing.

### CODE REVIEW

File: src/circuit/circuit_builder.rs [hunk 1]
```diff
@@ -301,6 +349,107 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
...
+/// owned value from the channel.  It is furthermore possible for the scheduler to schedule
+/// operators that strongly prefer owned values last.
+///
+/// We capture ownership preferences at two levels.  First, each individual operator that consumes
+/// one or more streams exposes its preferences on a per-stream basis via an API method
+/// (e.g., `[UnaryOperator::input_preference]`).  The [`Circuit`] API allows the circuit builder
+/// to override these preferences when instantiating an operator, taking into account circuit topology
+/// and workload.  We express preference as a numeric value.
+///
+/// These preferences are associated with each edge in the circuit graph.  The schedulers we have
```
Comment (mbudiu-vmw) [inline src/circuit/circuit_builder.rs:379]: I don't think this comment belongs here. The documentation should say what the objects do, not what the state of development is. The latter will change and this comment will become stale.
If this comment belongs somewhere is in the schedulers module, and as as // comment

File: src/circuit/circuit_builder.rs [hunk 2]
```diff
@@ -301,6 +349,107 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
...
+    pub fn numeric(&self) -> usize {
+        self.0
+    }
+}
+
+impl Display for OwnershipPreference {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if *self == Self::dont_care() {
+            f.write_str("?")
+        } else if *self == Self::prefer_owned() {
```
Comment (mbudiu-vmw) [inline src/circuit/circuit_builder.rs:424]: could be >=?

File: src/circuit/circuit_builder.rs [hunk 3]
```diff
@@ -301,6 +349,107 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
...
+            f.write_str("prefer_owned")
+        } else if *self == Self::require_owned() {
+            f.write_str("require_owned")
+        } else {
+            write!(f, "preference({})", self.numeric())
+        }
+    }
+}
+
+/// An edge in a circuit graph represents a stream connecting two
```
Comment (mbudiu-vmw) [inline src/circuit/circuit_builder.rs:434]: "Edges represent scheduling dependencies. These are of two kinds: actual data dependencies through streams and scheduling dependencies induced by preference constraints."

Comment (mbudiu-vmw) [reply src/circuit/circuit_builder.rs:434]: Let's call this Dependency, not Edge.

File: src/circuit/circuit_builder.rs [hunk 4]
```diff
@@ -301,6 +349,107 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
...
+/// operators or a dependency (i.e., a requirement that one operator
+/// must be evaluated before the other even if they are not connected
+/// by a stream).
+pub(super) struct Edge {
```
Comment (mbudiu-vmw) [inline src/circuit/circuit_builder.rs:438]: This should have some convenience methods, e.g., is_stream_dependency.

File: src/circuit/schedule/mod.rs [hunk 5]
```diff
@@ -130,3 +142,109 @@ where
...
+    /// This scheduling constraint can in turn be enforced by adding a dependency edge from
+    /// all other successors to the strong successor node to the circuit graph.  This function
+    /// computes the set of dependency edges needed to enforce all such constraints in the
+    /// circuit.
+    ///
+    /// # Caveat
+    ///
+    /// The additional edges computed by this function can introduce cycles to the circuit
+    /// graph making it unschedulable.  Schedulers must check for cycles before adding these
+    /// constraints and either fail or drop some of the constraints to eliminate cycles.
```
Comment (mbudiu-vmw) [inline src/circuit/schedule/mod.rs:187]: I assume that hard constraints cannot be dropped

**Review by mbudiu-vmw — APPROVED** [review]

File: src/circuit/operator/inspect.rs [hunk 6]
```diff
@@ -1,9 +1,48 @@
...
-use crate::circuit::operator_traits::{Operator, SinkOperator};
+use crate::circuit::{
+    operator_traits::{Operator, SinkOperator},
+    Circuit, Stream,
+};
 use std::{borrow::Cow, marker::PhantomData};
 
+/// Extenstion trait for [`Stream`].
+pub trait InspectTrait<D> {
+    /// Apply [`Inspect`] operator to `self`.
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/operator/inspect.rs:12]: I am going to just remove these extension traits.  They are in the same crate as `Stream`, so not really needed.

File: src/circuit/operator/integrate.rs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    /// after the current clock cycle.  The second stream contains the value of the
+    /// integral at the previous clock cycle, i.e., the sum of all inputs except
+    /// the last one.  The latter can equivalently be obtained by applying the delay
+    /// operator [`Z1`] to the integral, but this function avoids the extra storage
+    /// overhead and is the preferred way to perform delayed integration.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use dbsp::circuit::{
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/operator/integrate.rs:25]: These lines will be hidden in the generated documentation.

File: src/circuit/operator/integrate.rs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    /// ```
+    fn integrate_delayed(&self) -> Stream<C, D> {
+        self.integrate_core().1
+    }
+}
+
+impl<P, D> IntegrateTrait<Circuit<P>, D> for Stream<Circuit<P>, D>
+where
+    P: Clone + 'static,
+    D: Add<Output = D> + AddByRef + AddAssignByRef + Clone + HasZero + 'static,
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/operator/integrate.rs:89]: This trait bound precisely captures the requirements of this impl. There's no reason to strengthen it.

[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply src/circuit/operator/integrate.rs:89]: To make it easier to read and understand

File: src/circuit/operator/nested_source.rs [hunk 9]
```diff
@@ -55,9 +55,10 @@ where
     }
 
     fn clock_start(&mut self) {
-        self.val = unsafe { self.input_stream.get() }
-            .clone()
-            .unwrap_or_default();
+        self.val = match unsafe { self.input_stream.try_take() }.unwrap_or_default() {
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/operator/nested_source.rs:58]: This is a hack.  I am working to fix it.  Right now `None` can occur if the `start_clock` is from the parent scope.

File: src/circuit/schedule/mod.rs [hunk 10]
```diff
@@ -11,6 +11,14 @@ pub use dynamic_scheduler::DynamicScheduler;
...
 #[derive(Debug)]
 pub enum Error {
+    /// `origin` node has more than one strong successors who insist on consuming its output by value
+    /// (`OwnershipPreference::require_owned` or higher).
+    OwnershipConflict {
+        origin: GlobalNodeId,
+        consumers: Vec<GlobalNodeId>,
+    },
+    /// Ownership constraints introduce a cycle in the circuit graph.
+    CyclicCircuit { node_id: GlobalNodeId },
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/schedule/mod.rs:21]: Id of a node that is part of a cycle.  Ideally we'd return the whole cycle (preferably the shortest one).

File: src/circuit/circuit_builder.rs [hunk 11]
```diff
@@ -88,6 +92,11 @@ impl<C, D> Stream<C, D> {
     pub fn origin_node_id(&self) -> &GlobalNodeId {
         &self.origin_node_id
     }
+
+    /// Reference to the circuit the stream belongs to.
+    pub fn circuit(&self) -> &C {
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/circuit_builder.rs:97]: Is this really ambiguous without "parent".  This returns the circuit that the stream belongs to.  Yes, if the source node is not `None` then it is a node in the same circuit.  I used the parent/child terminology for nested circuits.

File: src/circuit/schedule/mod.rs (hunk 5 above)
Comment (ryzhyk) [reply src/circuit/schedule/mod.rs:187]: I'll clarify.  The current implementation models them as hard constraints but a different implementation may choose to relax them.

File: src/circuit/circuit_builder.rs (hunk 3 above)
Comment (ryzhyk) [reply src/circuit/circuit_builder.rs:434]: I don't see what's wrong with the comment or the name `Edge`. An edge is a stream or a dependency. Your terminology makes sense too, but it's not obviously more correct.

File: src/circuit/operator/plus.rs [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    ///     // Compute pairwise sums of values in the stream; the output stream will contain zeros.
+    ///     source1.plus(&source2).inspect(|n| assert_eq!(*n, 0));
+    /// })
+    /// .unwrap();
+    ///
+    /// # for _ in 0..5 {
+    /// #     root.step().unwrap();
+    /// # }
+    /// ```
+    fn plus(&self, other: &Stream<C, D>) -> Stream<C, D>;
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/operator/plus.rs:37]: true, add will make it consistent with other names.

File: src/circuit/circuit_builder.rs (hunk 3 above)
Comment (mbudiu-vmw) [reply src/circuit/circuit_builder.rs:434]: Edge is a much more abstract concept. These are edges that represent some specific relations in the circuit graph.

File: src/circuit/operator/plus.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/operator/plus.rs:37]: except it makes it easy to run into a name clash with `ops::Add`.  I'll keep the name `Plus`, it's too painful otherwise.

File: src/circuit/runtime.rs [hunk 13]
```diff
@@ -75,7 +75,7 @@ impl Runtime {
     ///
     ///     let root = Root::build(move |circuit| {
     ///         // Populate `circuit` with operators.
-    ///     });
+    ///     }).unwrap();
```
Comment (Kixiron) [inline src/circuit/runtime.rs:78]:
```suggestion
    ///     })
    ///     .unwrap();
```

File: src/circuit/schedule/dynamic_scheduler.rs [hunk 14]
```diff
@@ -215,17 +219,45 @@ impl Inner {
...
-        for (from, to) in circuit.edges().iter() {
-            successors.entry(*from).or_insert_with(Vec::new).push(*to);
-            predecessors.entry(*to).or_insert_with(Vec::new).push(*from);
+        for edge in circuit.edges().iter() {
+            if let Some(from) = edge.from {
+                successors
+                    .entry(from)
+                    .or_insert_with(Vec::new)
+                    .push(edge.to);
+                predecessors
```
Comment (Kixiron) [inline src/circuit/schedule/dynamic_scheduler.rs:250]:
```suggestion
                    .push(edge.to);

                predecessors
```

File: src/circuit/schedule/mod.rs [hunk 15]
```diff
@@ -130,3 +142,109 @@ where
         self.scheduler.step(circuit)
     }
 }
+
+mod util {
+    //! Some useful tools for developing schedulers.
```
Comment (Kixiron) [inline src/circuit/schedule/mod.rs:147]:
```suggestion
/// Some useful tools for developing schedulers.
mod util {
```

File: src/circuit/schedule/mod.rs [hunk 16]
```diff
@@ -130,3 +142,109 @@ where
...
+            if strong_successors.len() > 1 {
+                return Err(Error::OwnershipConflict {
+                    origin,
+                    consumers: strong_successors
+                        .into_iter()
+                        .map(|(_, (suc, _))| GlobalNodeId::child_of(circuit, *suc))
+                        .collect(),
+                });
+            };
+            // No strong successors -- nothing to do for this node.
```
Comment (Kixiron) [inline src/circuit/schedule/mod.rs:233]:
```suggestion
            };

            // No strong successors -- nothing to do for this node.
```

**Review by Kixiron — APPROVED** [review]

File: src/circuit/operator/inspect.rs [hunk 17]
```diff
@@ -1,9 +1,41 @@
...
-use crate::circuit::operator_traits::{Operator, SinkOperator};
+use crate::circuit::{
+    operator_traits::{Operator, SinkOperator},
+    Circuit, Stream,
+};
 use std::{borrow::Cow, marker::PhantomData};
 
+impl<P, D> Stream<Circuit<P>, D>
+where
+    D: Clone + 'static,
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/operator/inspect.rs:12]: I think at this point it comes down to using `Cow` to wrap values read from streams (regardless of an operator). We don't truly need `Cow` as we never do actual copy-on-write, so we could define our own `enum` instead that wouldn't require `Clone`. It would have to be another PR though and doesn't seem very pressing, as we generally want `Clone` for types inside the dataflow.  But yeah, shouldn't be strictly necessary any longer.

File: src/circuit/circuit_builder.rs [hunk 18]
```diff
@@ -301,6 +338,120 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
...
+/// and workload.  We express preference as a numeric value.
+///
+/// These preferences are associated with each edge in the circuit graph.  The schedulers we have
+/// built so far implement a limited form of ownership-aware scheduling.  They only consider strong
+/// preferences (`OwnershipPreference::require_owned` and stronger) and model them internally as hard
+/// constraints that must be satisfied for the circuit to be schedulable.  Weaker preferences are
+/// ignored.
+#[repr(transparent)]
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct OwnershipPreference(usize);
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/circuit_builder.rs:375]: 0 is a natural choice for expressing the "don't care" preference.  This type only plays a role during circuit construction, so no need to micro-optimize it.

File: src/circuit/circuit_builder.rs [hunk 19]
```diff
@@ -301,6 +338,120 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
...
+    pub from: Option<NodeId>,
+    /// Destination node.
+    pub to: NodeId,
+    /// Origin node that generates the stream.  If the origin belongs
+    /// to the local circuit, this is just the full path to the `from`
+    /// node.
+    pub origin: GlobalNodeId,
+    /// Ownership preference associated with the consumer of this
+    /// stream or `None` if this is a dependency edge.
+    pub ownership_preference: Option<OwnershipPreference>,
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/circuit_builder.rs:439]: A dependency edge means that there is an ordering requirement  between nodes that are _not_ connected by an actual stream (hence no ownership preference).

File: src/circuit/circuit_builder.rs [hunk 20]
```diff
@@ -301,6 +338,120 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
...
+            f.write_str("?")
+        } else if *self == Self::prefer_owned() {
+            f.write_str("prefer_owned")
+        } else if *self == Self::require_owned() {
+            f.write_str("require_owned")
+        } else {
+            write!(f, "preference({})", self.numeric())
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/circuit/circuit_builder.rs:421]: I've given this some thought.  Any kind of accurate cost model would be hard to capture and would certainly require more than a single numeric value.  E.g., it could be a function of the size of operator inputs. An individual operator lacks global visibility to express a meaningful preference. A scheduler that would respect such a model would have to solve an optimization problem, and would be pretty non-trivial both in terms of implementation and computational complexity. So instead I implement something simple that enables the really important optimization (avoid cloning the state stored in z^-1). Beyond that, the compiler or the user can make their own optimization decisions and communicate them to DBSP via dependecy edges.  The three constants are there because realistically these are the three meaningful preferences that operators can express.

**Comment by ryzhyk** [comment]
@Kixiron , thanks, I addressed most of your comments.
